frameworks/modular/module.c

```diff
@@ -0,0 +1,164 @@
+/* Copyright (c) 2017, ARM Limited. All rights reserved.
+ *
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include "module.h"
+
+#define MODULE_FRAMEWORK_VERSION 0x00010000UL
+ODP_SUBSYSTEM_DEFINE(module, "module framework", MODULE_FRAMEWORK_VERSION);
+
+/* Keep it simple, allow one registration session at a time. */
+static struct {
+	odp_rwlock_t lock;
+	odp_subsystem_t *subsystem;
+	odp_module_base_t *module;
+} registration = {
+	.lock = ODP_RWLOCK_UNLOCKED,
+	.subsystem = NULL,
+	.module = NULL,
+};
+
+#define REGISTRATION_SANITY_CHECK(subsystem, module)		\
+do {								\
+	if (subsystem == NULL || module == NULL)		\
+		return -ENOENT;					\
+								\
+	if (!list_node_detached(&module->list)) {		\
+		printf("module %s was already registered.\n",	\
+			module->name);				\
```

Comment:
In static link scenario, this will be prior to the main() entrance and ODP_DBG() ODP_LOG() facility are not ready yet?